x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportMlMemoryAction.java

```diff
@@ -0,0 +1,208 @@
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.node.DiscoveryNodeRole;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.settings.ClusterSettings;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.monitor.os.OsStats;
+import org.elasticsearch.persistent.PersistentTasksCustomMetadata;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.ml.action.MlMemoryAction;
+import org.elasticsearch.xpack.core.ml.action.MlMemoryAction.Response.MlMemoryStats;
+import org.elasticsearch.xpack.core.ml.action.TrainedModelCacheInfoAction;
+import org.elasticsearch.xpack.core.ml.action.TrainedModelCacheInfoAction.Response.CacheInfo;
+import org.elasticsearch.xpack.ml.job.NodeLoad;
+import org.elasticsearch.xpack.ml.job.NodeLoadDetector;
+import org.elasticsearch.xpack.ml.process.MlMemoryTracker;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+import static org.elasticsearch.xpack.ml.MachineLearning.MAX_MACHINE_MEMORY_PERCENT;
+import static org.elasticsearch.xpack.ml.MachineLearning.MAX_OPEN_JOBS_PER_NODE;
+import static org.elasticsearch.xpack.ml.MachineLearning.USE_AUTO_MACHINE_MEMORY_PERCENT;
+
+public class TransportMlMemoryAction extends TransportMasterNodeAction<MlMemoryAction.Request, MlMemoryAction.Response> {
+
+    private final Client client;
+    private final MlMemoryTracker memoryTracker;
+
+    @Inject
+    public TransportMlMemoryAction(
+        TransportService transportService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        ActionFilters actionFilters,
+        IndexNameExpressionResolver indexNameExpressionResolver,
+        Client client,
+        MlMemoryTracker memoryTracker
+    ) {
+        super(
+            MlMemoryAction.NAME,
+            transportService,
+            clusterService,
+            threadPool,
+            actionFilters,
+            MlMemoryAction.Request::new,
+            indexNameExpressionResolver,
+            MlMemoryAction.Response::new,
+            ThreadPool.Names.SAME
+        );
+        this.client = new OriginSettingClient(client, ML_ORIGIN);
+        this.memoryTracker = memoryTracker;
+    }
+
+    @Override
+    protected void masterOperation(
+        Task task,
+        MlMemoryAction.Request request,
+        ClusterState state,
+        ActionListener<MlMemoryAction.Response> listener
+    ) throws Exception {
+
+        ClusterSettings clusterSettings = clusterService.getClusterSettings();
+
+        // Resolve the node specification to some concrete nodes
+        String[] nodeIds = state.nodes().resolveNodes(request.getNodeId());
```

Comment:
Sneaky little method there. Resolve nodes is a deceptively complicated step!

Reply:
Yes, it's pretty cool.  I didn't know about the `ml:true` stuff until I started work on this PR.  That's been possible in the node info and node stats APIs since node roles were introduced.

Reply:
🤯